### PR TITLE
bpf: remove map migrations, add custom map flag

### LINKED ISSUE #29333: Remove 'map migration' mechanism
This supersedes https://github.com/cilium/cilium/issues/24013.

When an ELF is loaded and one or more of its maps are incompatible with a pinned map on bpffs, something needs to be done for ELF loading to proceed normally. Historically, we've opted for renaming the existing map pin with a `:pending` suffix, but this has the unfortunate downside of creating state to manage on bpffs.

As an alternative, I'm proposing to remove `{Start,Finalize}BPFFSMapMigration` altogether, and return a `commit()` closure from `bpf.LoadCollection()` instead of `finalize()` in `replaceDatapath()`. This would minimize changes in callers of `replaceDatapath()`.

As a follow-up, we can make `bpf.NewCollection` support custom pinning flags for e.g. per-endpoint maps and always-recreated maps like cilium_calls. `commit()` can be used to perform the pinning operations for those maps as well, e.g. for https://github.com/cilium/cilium/issues/29334.

### LINKED ISSUE #29334: Add custom BPF map pinning flag for 'always-recreated' maps
I added a hack in https://github.com/cilium/cilium/pull/28740/files#diff-5805cbfd909df5c9f2e6c668eb4ea2fee7eb5729a301933d7fd78b6e0e3d235dR104 to make sure `cilium_calls_*` maps are never picked up from a pin. Programs in cilium_calls are heavily dependent on each other, and making any modifications to an existing map is definitely unsafe.

This issue is for removing the special-case `bpf.RepinMap()` in `replaceDatapath` and replacing it with a custom map pin (making MapSpec.Pinning a bitfield) to signal to the ELF loader that MapSpec.Pinning needs to be cleared before loading, and that the map needs to be pinned by `commit()` (introduced in https://github.com/cilium/cilium/issues/29333).

Depends on https://github.com/cilium/cilium/issues/29333.

### PR DESCRIPTION
**Pull request by ti-mo**
Fixes: #29333
Fixes: #29334

Addressing both these issues in a single PR since they're relatively co-dependent.
See individual commits for details.

```release-note
Remove bpf map migration mechanism to minimize bpf file system operations during endpoint regeneration
```

### CODE REVIEW

**Review by lmb — APPROVED** [review]
Nice cleanup!

**Review by nebril — APPROVED** [review]
ci-structure LGTM

File: pkg/bpf/collection.go [hunk 1]
```diff
@@ -284,35 +284,68 @@ func LoadCollection(spec *ebpf.CollectionSpec, opts *CollectionOptions) (*ebpf.C
...
+
+		// Retry loading the Collection with necessary pinning flags removed.
+		coll, err = ebpf.NewCollectionWithOptions(spec, opts.CollectionOptions)
+	}
 
-		// Bump LogSize and retry if there's a truncated VerifierError.
+	// Try to obtain the full verifier log if it was truncated. Note that
+	// VerifierError is also returned if verification was successful but the
+	// buffer was too small.
+	attempts := 5
```
Comment (rgo3) [inline pkg/bpf/collection.go:311]: Nit: Feels a bit strange that we enter this code path even if there was no error at all. But of the top of my head I can only think of adding a label we can jump to early, and I'm not sure that makes it much nicer.

**Review by rgo3 — APPROVED** [review]
Nice change! One nit, but approving because it only impacts readability AFAICT.

File: pkg/bpf/collection.go (hunk 1 above)
Comment (ti-mo) [reply pkg/bpf/collection.go:311]: Yup, I've re-jigged this 4-5 times, and this was (to me) the most straightforward and understandable form, as awkward as it is. :/

**Review by dylandreimerink — APPROVED** [review]

**Comment by giorio94** [comment]
Updating the backport labels, as partial backports have been merged into v1.14 and v1.15.

**Pull request merged**